src/query/service/src/servers/http/v1/session/token_manager.rs

```diff
@@ -0,0 +1,218 @@
+            session_id: client_session_id.to_string(),
+            nonce: uuid::Uuid::new_v4().to_string(),
+            expire_at_in_secs: now + REFRESH_TOKEN_TTL,
+        };
+        let refresh_token = if let Some(old) = &old_token_pair {
+            old.refresh.clone()
+        } else {
+            claim.encode()
+        };
+        let refresh_token_hash = hash_token(refresh_token.as_bytes());
+        let token_info = QueryTokenInfo {
+            token_type: TokenType::Refresh,
+            parent: None,
+        };
+
+        meta_store
+            .upsert_token(
+                &tenant,
+                &refresh_token_hash,
+                token_info.clone(),
+                REFRESH_TOKEN_VALIDITY_IN_SECS + SESSION_TOKEN_VALIDITY_IN_SECS + TOKEN_TTL_DELAY,
+                false,
+            )
+            .await?;
+        if let Some(old) = old_token_pair {
+            meta_store
+                .upsert_token(
+                    &tenant,
+                    &old.refresh,
+                    token_info,
+                    REFRESH_TOKEN_VALIDITY_IN_SECS + TOKEN_DROP_DELAY,
+                    true,
+                )
+                .await?;
+        };
+        self.refresh_tokens
+            .write()
+            .put(refresh_token_hash.clone(), ());
+
+        claim.expire_at_in_secs = now + SESSION_TOKEN_VALIDITY_IN_SECS;
+        claim.nonce = uuid::Uuid::new_v4().to_string();
+
+        let session_token = claim.encode();
+        let session_token_hash = hash_token(session_token.as_bytes());
+        let token_info = QueryTokenInfo {
+            token_type: TokenType::Session,
+            parent: Some(refresh_token_hash.clone()),
+        };
+        meta_store
+            .upsert_token(
+                &tenant,
+                &session_token_hash,
+                token_info,
+                REFRESH_TOKEN_VALIDITY_IN_SECS + TOKEN_TTL_DELAY,
+                false,
+            )
+            .await?;
+        self.session_tokens.write().put(session_token_hash, ());
+
+        Ok((client_session_id, TokenPair {
+            refresh: refresh_token,
+            session: session_token,
+        }))
+    }
+
+    pub(crate) async fn verify_token(
+        self: &Arc<Self>,
+        token: &str,
+        token_type: TokenType,
+    ) -> Result<SessionClaim> {
+        let claim = SessionClaim::decode(token)?;
+        let now = unix_ts();
+        if now > claim.expire_at_in_secs {
+            return match token_type {
+                TokenType::Refresh => Err(ErrorCode::RefreshTokenExpired("refresh token expired")),
+                TokenType::Session => Err(ErrorCode::SessionTokenExpired("session token expired")),
+            };
+        }
+
+        let hash = hash_token(token.as_bytes());
+        let cache = match token_type {
+            TokenType::Refresh => &self.refresh_tokens,
+            TokenType::Session => &self.session_tokens,
+        };
+
+        if !cache.read().contains(&hash) {
```

Comment:
Can we return token hash or `session_id` to client and store token info at meta ? Because the information of the token is in plain text, user can construct it by themselves. It may be the same as the token generated by other users, which has security risks.

Reply:
session_id is fixed during session.  so can not `expired` as a token.